File: backend/app/schemas/delivery_order.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.models.delivery_order import DeliveryOrderStatus, DeliveryOrderType


# ── Brief models ─────────────────────────────────────────────────────────────

class CustomerBrief(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    first_name: str
    last_name: str
    company: str | None = None


class ContractBrief(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    contract_number: str
    status: str
    customer: CustomerBrief


class UserBrief(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    username: str
    full_name: str | None = None


# ── Checklist item schemas ──────────────────────────────────────────────────

class DeliveryChecklistItemCreate(BaseModel):
    item_group: str = Field(..., min_length=1, max_length=50)
    item_name: str = Field(..., min_length=1, max_length=200)
    is_passed: bool = True
    remark: str | None = None


class DeliveryChecklistItemOut(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    item_group: str
    item_name: str
    is_passed: bool
    remark: str | None = None
    created_at: datetime
    updated_at: datetime | None = None


# ── Delivery order schemas ──────────────────────────────────────────────────

class DeliveryOrderCreate(BaseModel):
    contract_id: int
    order_type: DeliveryOrderType = DeliveryOrderType.DELIVERY
    delivery_date: datetime
    delivery_address: str = Field(..., min_length=1)
    driver_name: str = Field(..., min_length=1, max_length=200)
    status: DeliveryOrderStatus = DeliveryOrderStatus.PENDING
    notes: str | None = None
    checklist_items: list[DeliveryChecklistItemCreate] | None = None


class DeliveryOrderUpdate(BaseModel):
    order_type: DeliveryOrderType | None = None
    delivery_date: datetime | None = None
    delivery_address: str | None = Field(default=None, min_length=1)
    driver_name: str | None = Field(default=None, min_length=1, max_length=200)
    status: DeliveryOrderStatus | None = None
    notes: str | None = None
    # When provided, fully replaces the existing checklist (delete + recreate)
    # — the checklist section on the form always resubmits its whole state.
    checklist_items: list[DeliveryChecklistItemCreate] | None = None


class DeliveryOrderResponse(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    do_no: str
    order_type: str
    contract: ContractBrief
    delivery_date: datetime
    delivery_address: str
    driver_name: str
    status: str
    notes: str | None = None
    checklist_items: list[DeliveryChecklistItemOut] = []
    creator: UserBrief | None = None
    created_at: datetime
    updated_at: datetime | None = None


class DeliveryOrderListResponse(BaseModel):
    items: list[DeliveryOrderResponse]
    total: int
    page: int
    page_size: int
    pages: int

File: frontend/src/types/deliveryOrder.ts
export type DeliveryOrderStatus = 'pending' | 'in_transit' | 'delivered' | 'cancelled'
export type DeliveryOrderType = 'delivery' | 'return'

export interface CustomerBrief {
  id: number
  first_name: string
  last_name: string
  company: string | null
}

export interface ContractBrief {
  id: number
  contract_number: string
  status: string
  customer: CustomerBrief
}

export interface UserBrief {
  id: number
  username: string
  full_name: string | null
}

export interface DeliveryChecklistItem {
  id: number
  item_group: string
  item_name: string
  is_passed: boolean
  remark: string | null
  created_at: string
  updated_at: string | null
}

// What the form sends — no `id`, since checklist updates are a full
// delete-and-recreate rather than a per-row diff (see the backend service).
export interface DeliveryChecklistItemInput {
  item_group: string
  item_name: string
  is_passed: boolean
  remark?: string | null
}

export interface DeliveryOrder {
  id: number
  do_no: string
  order_type: DeliveryOrderType
  contract: ContractBrief
  delivery_date: string
  delivery_address: string
  driver_name: string
  status: DeliveryOrderStatus
  notes: string | null
  checklist_items: DeliveryChecklistItem[]
  creator: UserBrief | null
  created_at: string
  updated_at: string | null
}

export interface DeliveryOrderListResponse {
  items: DeliveryOrder[]
  total: number
  page: number
  page_size: number
  pages: number
}

export interface DeliveryOrderCreate {
  contract_id: number
  order_type?: DeliveryOrderType
  delivery_date: string
  delivery_address: string
  driver_name: string
  status?: DeliveryOrderStatus
  notes?: string
  checklist_items?: DeliveryChecklistItemInput[]
}

export interface DeliveryOrderUpdate {
  order_type?: DeliveryOrderType
  delivery_date?: string
  delivery_address?: string
  driver_name?: string
  status?: DeliveryOrderStatus
  notes?: string
  checklist_items?: DeliveryChecklistItemInput[]
}

export interface DeliveryOrderListParams {
  q?: string
  status?: DeliveryOrderStatus
  contract_id?: number
  page?: number
  page_size?: number
}

// Payload handed from the rental contract detail view to the delivery order
// form via router state, for the Contract -> Delivery Order hand-off.
export interface DeliveryOrderConversionPrefill {
  contract_id: number
  contract_number: string
  customer_name: string
  forklift_label: string | null
  delivery_address: string | null
  order_type: DeliveryOrderType
  // The date to prefill into the form's own `delivery_date` field — the
  // contract's start_date for an outbound delivery, its end_date for a
  // return, decided by the caller before navigating here.
  delivery_date: string
}
